File: app/src/main/java/com/ourcookbook/domain/model/Recipe.kt
package com.ourcookbook.domain.model

import java.time.Instant
import java.util.UUID

/**
 * Domain model for Recipe
 * Represents a complete recipe with all metadata
 * 
 * This is the core domain model for recipes in the Cookbook application.
 * It contains all the essential information about a recipe including
 * ingredients, instructions, metadata, and synchronization information.
 */
data class Recipe(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val description: String? = null,
    val category: String,
    val ingredients: List<Ingredient> = emptyList(),
    val instructions: List<String> = emptyList(),
    val servingSize: Int? = null,
    val prepTime: Int? = null, // in minutes
    val cookTime: Int? = null, // in minutes
    val rating: Float? = null,
    val isFavorite: Boolean = false,
    val imageUrl: String? = null,
    val notes: String? = null,
    val source: String? = null,
    val tags: List<String> = emptyList(),
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
    val versionVector: VersionVector = VersionVector(),
    val checksum: String = "",
    val deviceId: String = ""
) {
    // Validate recipe has required fields
    fun isValid(): Boolean {
        return title.isNotBlank() && 
               category.isNotBlank() &&
               ingredients.isNotEmpty() &&
               instructions.isNotEmpty()
    }
    
    // Calculate total time (prep + cook)
    val totalTime: Int? get() {
        return when {
            prepTime == null && cookTime == null -> null
            prepTime == null -> cookTime
            cookTime == null -> prepTime
            else -> prepTime + cookTime
        }
    }
    
    // Check if recipe has an image
    val hasImage: Boolean get() = imageUrl != null && imageUrl.isNotBlank()
    
    // Get all ingredient names as a comma-separated string
    val ingredientNames: String get() = ingredients.joinToString(", ") { it.name }
    
    // Get formatted instructions as a single string
    val formattedInstructions: String get() = instructions.joinToString("\n\n")
    
    // Create a copy with updated timestamp and version
    fun withUpdate(deviceId: String): Recipe {
        return this.copy(
            versionVector = versionVector.increment(deviceId),
            updatedAt = Instant.now()
        )
    }
    
    companion object {
        fun create(
            title: String,
            category: String,
            ingredients: List<Ingredient> = emptyList(),
            instructions: List<String> = emptyList(),
            description: String? = null,
            servingSize: Int? = null,
            prepTime: Int? = null,
            cookTime: Int? = null,
            deviceId: String = ""
        ): Recipe {
            return Recipe(
                title = title,
                description = description,
                category = category,
                ingredients = ingredients,
                instructions = instructions,
                servingSize = servingSize,
                prepTime = prepTime,
                cookTime = cookTime,
                deviceId = deviceId
            )
        }
    }
}

File: webapp/src/types/recipe.ts
// Recipe — core domain model. Ported from Recipe.kt; field names match the
// Kotlin app so Drive-synced files stay backwards-compatible.

import type { Ingredient } from './ingredient'
import type { VersionVector } from './versionVector'
import { createVersionVector, incrementVersionVector } from './versionVector'

export interface Recipe {
  id: string
  title: string
  description: string | null
  category: string
  ingredients: Ingredient[]
  instructions: string[]
  servingSize: number | null
  prepTime: number | null // minutes
  cookTime: number | null // minutes
  rating: number | null
  isFavorite: boolean
  imageUrl: string | null
  notes: string | null
  source: string | null
  tags: string[]
  createdAt: string // ISO
  updatedAt: string // ISO
  versionVector: VersionVector
  checksum: string
  deviceId: string
}

export interface RecipeInput {
  title: string
  description?: string | null
  category: string
  ingredients?: Ingredient[]
  instructions?: string[]
  servingSize?: number | null
  prepTime?: number | null
  cookTime?: number | null
  rating?: number | null
  imageUrl?: string | null
  notes?: string | null
  source?: string | null
  tags?: string[]
  isFavorite?: boolean
}

export function createRecipe(input: RecipeInput, deviceId = ''): Recipe {
  const now = new Date().toISOString()
  return {
    id: crypto.randomUUID(),
    title: input.title,
    description: input.description ?? null,
    category: input.category,
    ingredients: input.ingredients ?? [],
    instructions: input.instructions ?? [],
    servingSize: input.servingSize ?? null,
    prepTime: input.prepTime ?? null,
    cookTime: input.cookTime ?? null,
    rating: input.rating ?? null,
    isFavorite: input.isFavorite ?? false,
    imageUrl: input.imageUrl ?? null,
    notes: input.notes ?? null,
    source: input.source ?? null,
    tags: input.tags ?? [],
    createdAt: now,
    updatedAt: now,
    versionVector: createVersionVector(deviceId),
    checksum: '',
    deviceId,
  }
}

export function withRecipeUpdate(recipe: Recipe, deviceId: string): Recipe {
  return {
    ...recipe,
    updatedAt: new Date().toISOString(),
    versionVector: incrementVersionVector(recipe.versionVector, deviceId),
  }
}

export function totalRecipeTime(r: Recipe): number | null {
  if (r.prepTime == null && r.cookTime == null) return null
  if (r.prepTime == null) return r.cookTime
  if (r.cookTime == null) return r.prepTime
  return r.prepTime + r.cookTime
}
